### PR TITLE
Workspace folder picker command (for #32936)

### LINKED ISSUE #32936: We should have a workspace folder picker API in the Extension host
Extensions now need to scope certain actions (like settings changes) to a workspace folder. In a multi workspace folder setup the need to ask the user which folder to use. To make such a picker the same for all extensions we should offer API for this.

### PR DESCRIPTION
**Pull request by bpasero**
fixes https://github.com/Microsoft/vscode/issues/32936

Introduces a simple new command `vscode.pickWorkspace` to pick a `WorkspaceFolder` via quick pick.

### CODE REVIEW

**Comment by bpasero** [comment]
@jrieken 

> What does this return? It must be triple-equal to the things we have in workspace.workspaceFolders. Also, why not have this as real API? Having the logic in the core shouldn't mean we cannot have a function for this

good point about the return type and identity, that is addressed in this change.

I rethought about having real API for this because I somewhat assume that this is a corner case where people might need it. If we want to go away from introducing more commands, I can also restore the proper API again and call into the command.

**Comment by jrieken** [comment]
> that this is a corner case where people might need it.

Fair enough, I let @dbaeumer decide if a command is OK for him

**Comment by jrieken** [comment]
Oh, and the test actually doesn't pass: https://travis-ci.org/Microsoft/vscode/jobs/277395186#L4159

**Comment by dbaeumer** [comment]
I always prefer explicit API over commands. The reason is:

- easier to discover
- I have code complete
- one place of documentation
- easier to keep compatible since we have compilers that check signatures. 

But that might be me :-)

**Comment by bpasero** [comment]
@jrieken @dbaeumer fair enough, I provide this API now (while still using the command internally):

```typescript
/**
 * Shows a selection list of [workspace folders](#workspace.workspaceFolders) to pick from.
 * Returns `undefined` if no folder is open.
 *
 * @param options Configures the behavior of the workspace folder list.
 * @return A promise that resolves to the workspace folder or `undefined`.
 */
export function showWorkspaceFolderPick(options?: WorkspaceFolderPickOptions): Thenable<WorkspaceFolder | undefined>;

/**
 * Options to configure the behaviour of the [workspace folder](#WorkspaceFolder) pick UI.
 */
export interface WorkspaceFolderPickOptions {

	/**
	 * An optional string to show as place holder in the input box to guide the user what to pick on.
	 */
	placeHolder?: string;

	/**
	 * Set to `true` to keep the picker open when focus moves to another part of the editor or to another window.
	 */
	ignoreFocusOut?: boolean;
}
```

File: src/vs/vscode.proposed.d.ts [hunk 1]
```diff
@@ -23,6 +23,31 @@ declare module 'vscode' {
...
+		 * @param options Configures the behavior of the workspace folder list.
+		 * @return A promise that resolves to the workspace folder or `undefined`.
+		 */
+		export function showWorkspaceFolderPick(options?: WorkspaceFolderPickOptions): Thenable<WorkspaceFolder | undefined>;
+	}
+
+	/**
+	 * Options to configure the behaviour of the [workspace folder](#WorkspaceFolder) pick UI.
+	 */
+	export interface WorkspaceFolderPickOptions {
```
Comment (jrieken) [inline src/vs/vscode.proposed.d.ts]: Don't define types inside a namespace but globally

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/vscode.proposed.d.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/vscode.proposed.d.ts]: @jrieken isn't that globally? It is outside the `window` namespace.

Comment (jrieken) [reply src/vs/vscode.proposed.d.ts]: Oh, sorry. Didn't see that closing-curly

**Review by jrieken — APPROVED** [review]

**Pull request merged**